src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.DivRem.cs

```diff
@@ -0,0 +1,360 @@
+            Debug.Assert(right.Length >= 1);
+            Debug.Assert(left.Length >= right.Length);
+
+            // Same as above, but only returning the quotient.
+
+            uint[] quotient = new uint[left.Length - right.Length + 1];
+
+            fixed (uint* l = left, r = right, q = quotient)
+            {
+                Divide(l, left.Length,
+                       r, right.Length,
+                       q, quotient.Length,
+                       null, 0);
+            }
+
+            return quotient;
+        }
+
+        [SecuritySafeCritical]
+        public unsafe static uint[] Remainder(uint[] left, uint[] right)
+        {
+            Debug.Assert(left != null);
+            Debug.Assert(right != null);
+            Debug.Assert(left.Length >= 1);
+            Debug.Assert(right.Length >= 1);
+            Debug.Assert(left.Length >= right.Length);
+
+            // Same as above, but only returning the remainder.
+
+            uint[] quotient = new uint[left.Length - right.Length + 1];
+            uint[] remainder = new uint[right.Length];
+
+            fixed (uint* l = left, r = right, q = quotient, e = remainder)
+            {
+                Divide(l, left.Length,
+                       r, right.Length,
+                       q, quotient.Length,
+                       e, remainder.Length);
+            }
+
+            return remainder;
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static void Divide(uint* left, int leftLength,
+                                          uint* right, int rightLength,
+                                          uint* quotient, int quotientLength,
+                                          uint* remainder, int remainderLength)
+        {
+            Debug.Assert(leftLength >= 1);
+            Debug.Assert(rightLength >= 1);
+            Debug.Assert(leftLength >= rightLength);
+            Debug.Assert(quotientLength == leftLength - rightLength + 1);
+            Debug.Assert(remainderLength == rightLength || remainderLength == 0);
+
+            // Executes the "grammar-school" algorithm for computing q = a / b.
+            // Before calculating q_i, we get more bits into the highest bit
+            // block of the divisor. Thus, guessing digits of the quotient
+            // will be more precise. Additionally we'll get r = a % b.
+
+            int dividendLength = leftLength + 1;
+            int divisorLength = rightLength;
+            fixed (uint* dividend = new uint[dividendLength],
+                         divisor = new uint[divisorLength],
+                         guess = new uint[divisorLength])
+            {
+                // This will create private copies of left and right, so we can
+                // modify the actual values of the dividend during computation
+                // without breaking immutability of the calling structure!
+                int shift = LeadingZeros(right[rightLength - 1]);
```

Comment:
Allocating extra space for the dividend is essential, since we need to subtract for every digit of the quotient a multiple of the divisor. The implementation of `BigIntegerBuilder` seems to do the same with calling `EnsureWritable`. And I've no idea how to do that without...

The extra space for the divisor spares the extra shifting logic within every digit calculation, since we can just shift the whole numbers once at the beginning and don't have to fumble around at every step. I can test with the server GC enabled, but since it is only one of four allocations (divisor vs. quotient, remainder and dividend) and beyond that the smaller one, I'm not that excited about it.

But, looking at the code again, I noticed that we could omit the allocation for the quotient, if we're only interested in the remainder. I'll do that anyway!